File: file_processing/worker.py
import asyncio
import logging
import socket

from temporalio.client import Client
from temporalio.worker import Worker

from activities import DEFAULT_TASK_QUEUE, download, process, upload
from workflows import FileProcessingWorkflow

# The host-specific task queue is simply this machine's hostname.
HOST_TASK_QUEUE = socket.gethostname()


async def main() -> None:
    logging.basicConfig(level=logging.INFO)

    client = await Client.connect("localhost:7233")

    # Worker 1 — handles workflows and the download activity (any host).
    default_worker = Worker(
        client,
        task_queue=DEFAULT_TASK_QUEUE,
        workflows=[FileProcessingWorkflow],
        activities=[download],
    )

    # Worker 2 — handles process and upload activities on this specific host.
    host_worker = Worker(
        client,
        task_queue=HOST_TASK_QUEUE,
        activities=[process, upload],
    )

    print(f"Default worker polling:    {DEFAULT_TASK_QUEUE!r}")
    print(f"Host worker polling:       {HOST_TASK_QUEUE!r}")
    print("Press Ctrl+C to stop.")

    # Run both workers concurrently.
    await asyncio.gather(default_worker.run(), host_worker.run())


if __name__ == "__main__":
    asyncio.run(main())

File: file_processing_ts/src/worker.ts
import { Worker } from '@temporalio/worker';
import { createDefaultActivities, createHostActivities } from './activities';
import { DEFAULT_TASK_QUEUE } from './workflows';
import * as os from 'os';

/**
 * Starts two workers:
 *
 *  1. Default worker  — polls DEFAULT_TASK_QUEUE; handles the workflow and
 *                       the `download` activity (any host can download).
 *
 *  2. Host worker     — polls a queue named after this machine's hostname;
 *                       handles `process` and `upload` (needs the local file).
 *
 * Using the hostname as the host-specific task queue name is what ties the
 * process and upload activities to the machine that ran the download.
 */
async function run() {
  const hostTaskQueue = os.hostname();

  const [defaultWorker, hostWorker] = await Promise.all([
    Worker.create({
      workflowsPath: require.resolve('./workflows'),
      activities: createDefaultActivities(),
      taskQueue: DEFAULT_TASK_QUEUE,
    }),
    Worker.create({
      activities: createHostActivities(),
      taskQueue: hostTaskQueue,
    }),
  ]);

  console.log(`Default worker polling: ${DEFAULT_TASK_QUEUE}`);
  console.log(`Host worker polling:    ${hostTaskQueue}`);
  console.log('Press Ctrl+C to stop.');

  await Promise.all([defaultWorker.run(), hostWorker.run()]);
}

run().catch((err) => {
  console.error(err);
  process.exit(1);
});
